openedx/core/djangoapps/verified_track_content/partition_scheme.py

```diff
@@ -0,0 +1,147 @@
+"""
+UserPartitionScheme for enrollment tracks.
+"""
+from courseware.masquerade import (  # pylint: disable=import-error
+    get_course_masquerade,
+    get_masquerading_group_info,
+    is_masquerading_as_specific_student,
+)
+from course_modes.models import CourseMode, EnrollmentMode
+from student.models import CourseEnrollment
+from opaque_keys.edx.keys import CourseKey
+from openedx.core.djangoapps.verified_track_content.models import VerifiedTrackCohortedCourse
+from xmodule.partitions.partitions import NoSuchUserPartitionGroupError, Group, UserPartition
+
+
+# EnrollmentMode IDs will be low integers (starting at 0), which is important so that they
+# do not overlap with Groups in the CohortUserPartition or the RandomUserPartitionScheme
+# (CMS' course_group_config uses a minimum value of 100 for all generated IDs).
+ENROLLMENT_GROUP_IDS = {}
+try:
+    for enrollment_mode in EnrollmentMode.objects.all():
+        ENROLLMENT_GROUP_IDS[enrollment_mode.mode_slug] = enrollment_mode.id
+except:  # pylint: disable=bare-except
+    pass
+
+if not ENROLLMENT_GROUP_IDS:
+    # The entries in this table are populated in a migration file, and migration files do not
+    # run as part of unit tests. Therefore, for tests hard-code the modes.
+    ENROLLMENT_GROUP_IDS = {
+        CourseMode.AUDIT: 1,
+        CourseMode.VERIFIED: 2,
+        CourseMode.PROFESSIONAL: 3,
+        CourseMode.NO_ID_PROFESSIONAL_MODE: 4,
+        CourseMode.CREDIT_MODE: 5,
+        CourseMode.HONOR: 6
+    }
+
+
+class EnrollmentTrackUserPartition(UserPartition):
+    """
+    Extends UserPartition to support dynamic groups pulled from the current course Enrollment tracks.
+    """
+
+    @property
+    def groups(self):
+        """
+        Return the groups (based on CourseModes) for the course associated with this
+        EnrollmentTrackUserPartition instance.
+        """
+        course_key = CourseKey.from_string(self.parameters["course_id"])
+
+        if is_course_using_cohort_instead(course_key):
+            return []
+
+        all_groups = []
+        for mode in CourseMode.modes_for_course(course_key, include_expired=True, only_selectable=False):
+            group = Group(ENROLLMENT_GROUP_IDS[mode.slug], unicode(mode.name))
+            all_groups.append(group)
+
+        return all_groups
```

Comment:
Nitty: Consider using a list comprehension syntax for readability, consolidating the above 6 lines of code.

```
return [
    Group(ENROLLMENT_GROUP_IDS[mode.slug], unicode(mode.name))
    for mode in CourseMode.modes_for_course(course_key, include_expired=True, only_selectable=False)
]
```